File: web/static/obfuscate.js
/**
 * @typedef {Object} ObfuscateOption
 * @property {string} shell - Specify the shell.
 * @property {string} varNamePrefix - Variable name prefix.
 * @property {number} sliceLength - Slice length.
 * @property {number} minVarNameLength - Minimum variable name length.
 * @property {boolean} useTempFile - Create a temporary file for execution.
 * @property {number} argLengthLimit - Argument length limit; exceeding this limit will force script execution through the creation of a temporary file.
 * @property {number} varCountLimit - Variable count limit; exceeding this limit will automatically set the slice length.
 */

/** 
 * @param {string} codes
 * @param {ObfuscateOption} option default {
        sliceLength: 5,
        minVarNameLength: 3,
        varNamePrefix: "_",
        useTempFile: false,
        argLengthLimit: 17200,
        varCountLimit: 20000,
    }
*/
async function obfuscateShellScript(codes, option = {}) {
    const defaultOption = {
        sliceLength: 5,
        minVarNameLength: 3,
        varNamePrefix: "_",
        useTempFile: false,
        argLengthLimit: 17200,
        varCountLimit: 20000,
    };
    let { shell, sliceLength, varNamePrefix, minVarNameLength, useTempFile, argLengthLimit, varCountLimit } = Object.assign({}, defaultOption, option);
    const lines = codes.split("\n");
    const shebang = [-1, ""];
    for (let i = 0; i < lines.length; i++) {
        const l = lines[i].trim();
        if (l !== "") {
            if (l.startsWith("#!")) {
                shebang[0] = i;
                shebang[1] = l;
            }
            break;
        }
    }
    if (shebang[0] > -1) {
        lines.splice(shebang[0], 1);
        codes = lines.join("\n");
    }
    if (["", null, undefined].includes(shell)) {
        if (shebang[0] > -1) {
            shell = "";
            const ss = shebang[1]
                .split(" ")
                .map(e => e.trim())
                .filter(e => e !== "");
            if (ss.length > 0) {
                if (ss.length === 1) {
                    shell = ss[0].replace("#!", "");
                } else {
                    shell = ss[1];
                }
            }
        } else {
            shell = "bash";
        }
    }
    if (shell === "") {
        shell = "bash";
    }
    /**
     * @param {string} s
     * @return {Promise<string>}
     */
    const b64encode = s => {
        return new Promise((resolve, reject) => {
            if (typeof Buffer === "function") {
                resolve(Buffer.from(s).toString("base64"));
                return;
            }
            const blob = new Blob([s], { type: "text/plain" });
            const reader = new FileReader();
            reader.onload = () => {
                resolve(reader.result.replace("data:text/plain;base64,", ""));
            };
            reader.onerror = reject;
            reader.readAsDataURL(blob);
        });
    };
    const chars = Array(26)
        .fill(null)
        .map((_, i) => {
            return String.fromCharCode(65 + i);
        })
        .concat(
            Array(26)
                .fill(null)
                .map((_, i) => {
                    return String.fromCharCode(97 + i);
                })
        );
    const genVarName = (length = 1) => {
        let v = "";
        for (let i = 0; i < length; i++) {
            v += chars[Math.floor(Math.random() * chars.length)];
        }
        return v;
    };
    const b64Codes = await b64encode(codes);
    const tempFile = varNamePrefix + genVarName(3);
    let evalCodes = "";
    // If the maximum limit on the length of command line arguments is exceeded, a temporary file is used
    useTempFile = useTempFile || b64Codes.length > argLengthLimit;
    if (useTempFile) {
        evalCodes = `trap "rm -f \\$${tempFile}" EXIT;${tempFile}=$(mktemp) || exit 1; echo ${b64Codes} | base64 -d > "$${tempFile}";${shell} "$${tempFile}" "$@";`;
    } else {
        evalCodes = `${shell} -c "$(base64 -d <<< "${b64Codes}")" ${shell} "$@"`;
    }
    // variable count
    let varCount = Math.ceil(evalCodes.length / sliceLength);
    // setting a maximum limit on the number of variables to prevent execution overflow
    if (varCount > varCountLimit) {
        varCount = varCountLimit;
        sliceLength = Math.ceil(evalCodes.length / varCount);
    }
    // variable name length
    let varNameLength = minVarNameLength;
    while (chars.length ** varNameLength < varCount) {
        varNameLength++;
    }
    // generate
    let tmp = evalCodes;
    let varNameSet = new Set();
    let varGroups = [];
    while (tmp.length > 0) {
        const value = tmp.slice(0, sliceLength);
        tmp = tmp.slice(sliceLength);
        let key = "";
        do {
            key = genVarName(varNameLength);
        } while (varNameSet.has(key));
        varNameSet.add(key);
        // prefix
        key = varNamePrefix + key;
        varGroups.push({ key, value });
    }

    let result = "";
    if (shebang[0] >= 0) {
        result += shebang[1] + "\n";
    }
    // Disrupt order
    for (const { key, value } of varGroups.map(e => e).sort(() => Math.random() - 0.5)) {
        result += `${key}='${value}';`;
    }
    result += `eval "${varGroups.map(e => "$" + e.key).join("")}";`;
    return {
        source: codes,
        result: result,
        useTempFile,
        varCount,
        sliceLength,
    };
}

File: shc/obfuscate.go
package shc

import (
	"encoding/base64"
	"fmt"
	"math"
	"math/rand"
	"sort"
	"time"
)

type ObfuscateOption struct {
	// Specify a shell that is automatically parsed and configured by default
	Shell string
	// Variable prefix, default "_"
	VarNamePrefix string
	// Slice length, default 5
	SliceLength int
	// Minimum variable length, default 3
	MinVarNameLength int
	// Run by creating a temporary file
	UseTempFile bool
	// Argument length limit; exceeding this limit will force script execution through the creation of a temporary file. default 17200
	ArgLengthLimit int
	// Variable count limit; exceeding this limit will automatically set the slice length. default 20000
	VarCountLimit int
}

type ObfuscateResult struct {
	Output string
	ObfuscateOption
}

func ObfuscateShellScript(script string, opt ObfuscateOption) *ObfuscateResult {
	if opt.SliceLength < 1 {
		opt.SliceLength = 5
	}
	if opt.MinVarNameLength < 1 {
		opt.MinVarNameLength = 3
	}
	if opt.ArgLengthLimit < 1 {
		opt.ArgLengthLimit = 17200
	}
	if opt.VarCountLimit < 1 {
		opt.VarCountLimit = 20000
	}
	if opt.VarNamePrefix == "" {
		opt.VarNamePrefix = "_"
	}
	shebang, shell, line := ParseShebang(script)
	if opt.Shell == "" {
		if line < 0 {
			opt.Shell = "bash"
		} else {
			opt.Shell = shell
		}
	}
	chars := "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
	r := rand.New(rand.NewSource(time.Now().UnixMilli()))
	genVarName := func(length int) string {
		bs := []byte{}
		for i := 0; i < length; i++ {
			bs = append(bs, chars[r.Intn(len(chars))])
		}
		return string(bs)
	}
	b64Codes := base64.StdEncoding.EncodeToString([]byte(script))
	tempFile := opt.VarNamePrefix + genVarName(3)
	evalCodes := ""
	// If the maximum limit on the length of command line arguments is exceeded, a temporary file is used
	opt.UseTempFile = opt.UseTempFile || len(b64Codes) > opt.ArgLengthLimit
	if opt.UseTempFile {
		evalCodes = fmt.Sprintf(`trap "rm -f \$%s" EXIT;%s=$(mktemp) || exit 1; echo %s | base64 -d > "$%s";%s "$%s" "$@";`, tempFile, tempFile, b64Codes, tempFile, opt.Shell, tempFile)
	} else {
		evalCodes = fmt.Sprintf(`%s -c "$(base64 -d <<< "%s")" %s "$@"`, opt.Shell, b64Codes, opt.Shell)
	}
	// setting a maximum limit on the number of variables to prevent execution overflow
	varCount := int(math.Ceil(float64(len(evalCodes)) / float64(opt.SliceLength)))
	if varCount > opt.VarCountLimit {
		varCount = opt.VarCountLimit
		opt.SliceLength = int(math.Ceil(float64(len(evalCodes)) / float64(varCount)))
	}
	varNameLength := opt.MinVarNameLength
	for int(math.Pow(float64(len(chars)), float64(varNameLength))) < opt.VarCountLimit {
		varNameLength++
	}
	// generate
	tmp := evalCodes
	varNameMap := make(map[string]struct{})
	varGroups := [][]string{}
	for len(tmp) > 0 {
		sliceLength := min(opt.SliceLength, len(tmp))
		v := tmp[0:sliceLength]
		tmp = tmp[sliceLength:]
		k := ""
		for {
			k = genVarName(varNameLength)
			if _, ok := varNameMap[k]; !ok {
				break
			}
		}
		varNameMap[k] = struct{}{}
		k = opt.VarNamePrefix + k
		varGroups = append(varGroups, []string{k, v})
	}
	result := ""
	if line > -1 {
		result = shebang + "\n"
	}
	varNames := make([]string, 0, len(varGroups))
	for _, v := range varGroups {
		varNames = append(varNames, v[0])
	}
	// Disrupt order
	sort.Slice(varGroups, func(i, j int) bool {
		return rand.Intn(2) == 0
	})
	for _, v := range varGroups {
		result += fmt.Sprintf(`%s='%s';`, v[0], v[1])
	}
	result += `eval "`
	for _, v := range varNames {
		result += fmt.Sprintf(`$%s`, v)
	}
	result += `";`
	return &ObfuscateResult{
		Output:          result,
		ObfuscateOption: opt,
	}
}
